### PR TITLE
Adjustments to the Result and Logger Interfaces

### PR DESCRIPTION
**Pull request by justing-bq**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: qldbdriver/logger.go [hunk 1]
```diff
@@ -58,18 +58,23 @@ func (qldbLogger *qldbLogger) logf(verbosityLevel LogLevel, message string, args
...
+type defaultLogger struct {
+	// The verbosity level of the logs that the logger should receive.
+	verbosity LogLevel
+}
 
 // Log the message using the built-in Golang logging package.
-func (logger defaultLogger) Log(message string) {
-	log.Println(message)
+func (logger *defaultLogger) Log(message string, verbosity LogLevel) {
+	if verbosity <= logger.verbosity {
```
Comment (simonz-bq) [inline qldbdriver/logger.go:77]: The logger implementation itself is NOT what controls what comes through. There should be no changes here.

File: qldbdriver/logger.go [hunk 2]
```diff
@@ -58,18 +58,23 @@ func (qldbLogger *qldbLogger) logf(verbosityLevel LogLevel, message string, args
...
-			qldbLogger.logger.Log(fmt.Sprintf(message, args...))
+			qldbLogger.logger.Log(fmt.Sprintf(message, args...), verbosityLevel)
 		}
 	}
 }
 
-type defaultLogger struct{}
+type defaultLogger struct {
+	// The verbosity level of the logs that the logger should receive.
+	verbosity LogLevel
```
Comment (simonz-bq) [inline qldbdriver/logger.go:72]: This is not needed.

File: qldbdriver/qldbdriver.go [hunk 3]
```diff
@@ -71,6 +71,11 @@ func New(ledgerName string, qldbSession *qldbsession.QLDBSession, fns ...func(*D
 		return nil, &qldbDriverError{"MaxConcurrentTransactions must be 1 or greater."}
 	}
 
+	// Ensure logger's internal verbosity is consistent with the option's verbosity for when defaultLogger is used.
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:74]: This is not how the current logging works. This block of code is not needed.

File: qldbdriver/qldbdriver.go [hunk 4]
```diff
@@ -171,8 +176,9 @@ func (driver *QLDBDriver) GetTableNames(ctx context.Context) ([]string, error) {
 		if err != nil {
 			return nil, err
 		}
+
 		tableNames := make([]string, 0)
-		for result.Next(txn) {
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:175]: Why is this change here? Removing txn from Next breaks a very important concept.

File: qldbdriver/result.go [hunk 5]
```diff
@@ -19,8 +19,17 @@ import (
...
+type Result interface {
+	Next() bool
+	GetCurrentData() []byte
+	GetConsumedIOs() *IOUsage
+	GetTimingInformation() *TimingInformation
+	Err() error
+}
+
+// qldbResult is a cursor over a result set from a QLDB statement.
+type qldbResult struct {
```
Comment (simonz-bq) [inline qldbdriver/result.go:32]: just call it result, lowercase.

File: qldbdriver/result.go [hunk 6]
```diff
@@ -19,8 +19,17 @@ import (
 	"github.com/aws/aws-sdk-go/service/qldbsession"
 )
 
-// Result is a cursor over a result set from a QLDB statement.
-type Result struct {
+// Result is an interface for a Result Set cursor.
+type Result interface {
```
Comment (simonz-bq) [inline qldbdriver/result.go:23]: Please do not change this API.

File: qldbdriver/result.go [hunk 7]
```diff
@@ -151,6 +161,12 @@ func (result *BufferedResult) GetTimingInformation() *TimingInformation {
 	return result.metrics.timingInformation
 }
 
+// Err is defined so BufferedResult will formally implement the Result interface.
+// Currently there are no error cases so we just return nil.
+func (result *BufferedResult) Err() error {
```
Comment (simonz-bq) [inline qldbdriver/result.go:166]: We shouldn't do this. They aren't supposed to have the same API. Add a BufferedResult interface as well.

File: qldbdriver/transaction.go [hunk 8]
```diff
@@ -25,9 +25,9 @@ import (
 // Transaction represents an active QLDB transaction.
 type Transaction interface {
 	// Execute a statement with any parameters within this transaction.
-	Execute(statement string, parameters ...interface{}) (*Result, error)
-	// Buffer a Result into a BufferedResult to use outside the context of this transaction.
-	BufferResult(result *Result) (*BufferedResult, error)
+	Execute(statement string, parameters ...interface{}) (Result, error)
+	// Buffer a qldbResult into a BufferedResult to use outside the context of this transaction.
```
Comment (simonz-bq) [inline qldbdriver/transaction.go:29]: This docstring should not be changed. The user of the docs has no idea, and shouldn't, what a "qldbResult" is.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: qldbdriver/qldbdriver.go (hunk 4 above)
Comment (justing-bq) [reply qldbdriver/qldbdriver.go:175]: Because the implementation of `result.Next(txn)` doesn't actually do anything with `txn` beyond pass it into the recursive call of `result.Next(txn)`.
And also because BufferedResult has a `Next() bool` method defined allowing that method to be part of the Result interface.

File: qldbdriver/result.go (hunk 6 above)
Comment (justing-bq) [reply qldbdriver/result.go:23]: This interface is an addition to the code not a modification.
The method signatures in here are the methods shared between `result` and `BufferedResult` and also what is used by the tests dealing with a `Result` interface.

File: qldbdriver/logger.go (hunk 1 above)
Comment (justing-bq) [reply qldbdriver/logger.go]: Addressed.

File: qldbdriver/logger.go (hunk 2 above)
Comment (justing-bq) [reply qldbdriver/logger.go]: Addressed.

File: qldbdriver/qldbdriver.go (hunk 3 above)
Comment (justing-bq) [reply qldbdriver/qldbdriver.go]: Addressed.

File: qldbdriver/result.go (hunk 5 above)
Comment (justing-bq) [reply qldbdriver/result.go]: Addressed.

File: qldbdriver/qldbdriver.go (hunk 4 above)
Comment (justing-bq) [reply qldbdriver/qldbdriver.go:175]: Added txn back to Next() and updated Result interface accordingly.

File: qldbdriver/transaction.go [hunk 9]
```diff
@@ -107,20 +107,26 @@ type transactionExecutor struct {
...
 // Execute a statement with any parameters within this transaction.
-func (executor *transactionExecutor) Execute(statement string, parameters ...interface{}) (*Result, error) {
+func (executor *transactionExecutor) Execute(statement string, parameters ...interface{}) (Result, error) {
 	return executor.txn.execute(executor.ctx, statement, parameters...)
 }
 
-// Buffer a Result into a BufferedResult to use outside the context of this transaction.
-func (executor *transactionExecutor) BufferResult(result *Result) (*BufferedResult, error) {
+// Buffer a result into a bufferedResult to use outside the context of this transaction.
+func (executor *transactionExecutor) BufferResult(res Result) (*bufferedResult, error) {
```
Comment (simonz-bq) [inline qldbdriver/transaction.go:115]: I think you forgit to add the BufferedResult interface.

We're returning a type we don't export here.

File: qldbdriver/transaction.go [hunk 10]
```diff
@@ -107,20 +107,26 @@ type transactionExecutor struct {
 }
 
 // Execute a statement with any parameters within this transaction.
-func (executor *transactionExecutor) Execute(statement string, parameters ...interface{}) (*Result, error) {
+func (executor *transactionExecutor) Execute(statement string, parameters ...interface{}) (Result, error) {
 	return executor.txn.execute(executor.ctx, statement, parameters...)
 }
 
-// Buffer a Result into a BufferedResult to use outside the context of this transaction.
```
Comment (simonz-bq) [inline qldbdriver/transaction.go:114]: This docstring doesn't need to change, as Result and BufferedResult are (or should now be) the interfaces which are consumed and returned.

File: qldbdriver/transaction.go [hunk 11]
```diff
@@ -107,20 +107,26 @@ type transactionExecutor struct {
...
 	}
-	if result.Err() != nil {
-		return nil, result.Err()
+	if res.Err() != nil {
+		return nil, res.Err()
 	}
-	return &BufferedResult{bufferedResults, 0, nil, result.metrics}, nil
+
+	var metrics *metrics = nil
+	if r, ok := res.(*result); ok {
```
Comment (simonz-bq) [inline qldbdriver/transaction.go:125]: Why is this block of code needed?

File: qldbdriver/result.go (hunk 6 above)
Comment (simonz-bq) [reply qldbdriver/result.go:23]: I was referring to how Next(txn) was changed, but it's been reverted so this isnt an issue anymore.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: qldbdriver/transaction.go (hunk 11 above)
Comment (justing-bq) [reply qldbdriver/transaction.go:125]: Because the passed in `res` is of interface type `Result` (not concrete type `result`) so this is needed to get the metrics from the underlying `result` object.

File: qldbdriver/transaction.go (hunk 9 above)
Comment (justing-bq) [reply qldbdriver/transaction.go]: Added BufferedResult interface.

File: qldbdriver/transaction.go (hunk 10 above)
Comment (justing-bq) [reply qldbdriver/transaction.go:114]: Changed comment back.

File: qldbdriver/transaction.go [hunk 12]
```diff
@@ -25,9 +25,9 @@ import (
 // Transaction represents an active QLDB transaction.
 type Transaction interface {
 	// Execute a statement with any parameters within this transaction.
-	Execute(statement string, parameters ...interface{}) (*Result, error)
-	// Buffer a Result into a BufferedResult to use outside the context of this transaction.
-	BufferResult(result *Result) (*BufferedResult, error)
+	Execute(statement string, parameters ...interface{}) (Result, error)
+	// Buffer a Result into a bufferedResult to use outside the context of this transaction.
+	BufferResult(res Result) (*bufferedResult, error)
```
Comment (simonz-bq) [inline qldbdriver/transaction.go]: This should return the interface.

File: qldbdriver/transaction.go [hunk 13]
```diff
@@ -39,7 +39,7 @@ type transaction struct {
 	commitHash   *qldbHash
 }
 
-func (txn *transaction) execute(ctx context.Context, statement string, parameters ...interface{}) (*Result, error) {
+func (txn *transaction) execute(ctx context.Context, statement string, parameters ...interface{}) (Result, error) {
```
Comment (simonz-bq) [inline qldbdriver/transaction.go:42]: Nit: I don't think the internal transaction needs to return Result instead of *result.

File: qldbdriver/transaction.go (hunk 11 above)
Comment (simonz-bq) [reply qldbdriver/transaction.go:125]: I see, but this is still type checking and casting when in this case we don't need to do this.

We can just change bufferedResult to take in the metrics individually after we get them from Result and change its functionality to return them individually.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: qldbdriver/transaction.go (hunk 11 above)
Comment (justing-bq) [reply qldbdriver/transaction.go]: Addressed.

File: qldbdriver/transaction.go (hunk 13 above)
Comment (justing-bq) [reply qldbdriver/transaction.go]: Addressed.

File: qldbdriver/transaction.go (hunk 8 above)
Comment (justing-bq) [reply qldbdriver/transaction.go]: Addressed.

File: qldbdriver/result.go [hunk 14]
```diff
@@ -19,8 +19,17 @@ import (
 	"github.com/aws/aws-sdk-go/service/qldbsession"
 )
 
-// Result is a cursor over a result set from a QLDB statement.
-type Result struct {
+// Result is an interface for a Result Set cursor.
```
Comment (simonz-bq) [inline qldbdriver/result.go:22]: This should keep the old, existing docstring.

File: qldbdriver/result.go [hunk 15]
```diff
@@ -59,71 +69,78 @@ func (result *Result) Next(txn Transaction) bool {
...
 }
 
 // Err returns an error if a previous call to Next has failed.
 // The returned error will be nil if the previous call to Next succeeded.
-func (result *Result) Err() error {
+func (result *result) Err() error {
 	return result.err
 }
 
-// BufferedResult is a cursor over a result set from a QLDB statement that is valid outside the context of a transaction.
```
Comment (simonz-bq) [inline qldbdriver/result.go:115]: Ditto here regarding the docstring. This is what the user says so it should keep the old docstring.

File: qldbdriver/result.go [hunk 16]
```diff
@@ -157,6 +174,11 @@ type IOUsage struct {
 	writeIOs *int64
 }
 
+// NewIOUsage creates a new instance of IOUsage.
```
Comment (simonz-bq) [inline qldbdriver/result.go:177]: Why do we need this and NewTimingInformation. I don't think we need them.

Even if we do, they should not be exported.

File: qldbdriver/result.go [hunk 17]
```diff
@@ -83,30 +84,27 @@ func (result *result) getNextPage() error {
...
 	if fetchPageResult.TimingInformation != nil {
-		*result.metrics.timingInformation.processingTimeMilliseconds += *fetchPageResult.TimingInformation.ProcessingTimeMilliseconds
+		*result.timingInfo.processingTimeMilliseconds += *fetchPageResult.TimingInformation.ProcessingTimeMilliseconds
 	}
 }
 
 // GetConsumedIOs returns the statement statistics for the current number of read IO requests that were consumed. The statistics are stateful.
 func (result *result) GetConsumedIOs() *IOUsage {
-	readIOs := *result.metrics.ioUsage.readIOs
-	writeIOs := *result.metrics.ioUsage.writeIOs
```
Comment (simonz-bq) [inline qldbdriver/result.go:88]: I believe this is done to prevent the pointer of the actual data type from being overwritten if the user reassigned what they got back.

Please revert this part for this and ProcessingTime.

File: qldbdriver/result.go (hunk 14 above)
Comment (justing-bq) [reply qldbdriver/result.go:22]: The `Result struct` got renamed to `result struct` which has the same docstring.
This new docstring is for the `Result interface`. Are you saying this interface should have the same docstring as the struct?

File: qldbdriver/result.go (hunk 15 above)
Comment (justing-bq) [reply qldbdriver/result.go:115]: Same question as above.

File: qldbdriver/result.go (hunk 16 above)
Comment (justing-bq) [reply qldbdriver/result.go:177]: Mostly just for convenience. Since the `IOUsage` and `TimingInformation` structs take in pointers to ints, it's easier to use these constructors than to have to do this:

```go
readIOs := 0
writeIOs := 0
ioUsage := &IOUsage{&readIOs, &writeIOs}
```

Anyway I've unexported the constructors.

Comment (justing-bq) [reply qldbdriver/result.go:177]: Mostly just for convenience. Since these structs take in pointers to ints, its easier to use these constructors than to have to do this kinda thing every time:

```go
readIOs := 0
writeIOs := 0
ioUsage := &IOUsage{&readIOs, &writeIOs}
```

Comment (justing-bq) [reply qldbdriver/result.go:177]: The constructors have been un-exported.

File: qldbdriver/result.go (hunk 17 above)
Comment (justing-bq) [reply qldbdriver/result.go:88]: Done.

File: qldbdriver/result.go (hunk 14 above)
Comment (justing-bq) [reply qldbdriver/result.go]: Addressed.

File: qldbdriver/result.go (hunk 15 above)
Comment (justing-bq) [reply qldbdriver/result.go:115]: Addressed.
